src/openvic-simulation/history/CountryHistory.cpp

```diff
@@ -0,0 +1,315 @@
+	}
+	Logger::info("Locked country history registry after registering ", country_histories.size(), " items");
+	locked = true;
+}
+
+bool CountryHistoryManager::is_locked() const {
+	return locked;
+}
+
+CountryHistory const* CountryHistoryManager::get_country_history(Country const* country, Date entry) const {
+	Date closest_entry;
+	auto country_registry = country_histories.find(country);
+
+	if (country_registry == country_histories.end()) {
+		Logger::error("Attempted to access history of undefined country ", country->get_identifier());
+		return nullptr;
+	}
+	
+	for (const auto& current : country_registry->second) {
+		if (current.first == entry) return &current.second;
+		if (current.first > entry) continue;
+		if (current.first > closest_entry && current.first < entry) closest_entry = current.first;
+	}
+
+	auto entry_registry = country_registry->second.find(closest_entry);
+	if (entry_registry == country_registry->second.end()) {
+		return &entry_registry->second;
+	}
+	/* warned about lack of entries earlier, return nullptr */
+	return nullptr;
+}
+
+inline CountryHistory const* CountryHistoryManager::get_country_history(Country const* country, Bookmark const* entry) const {
+	return get_country_history(country, entry->get_date());
+}
+
+inline bool CountryHistoryManager::_load_country_history_entry(GameManager& game_manager, std::string_view name, OpenVic::Date date, ast::NodeCPtr root) {
+	Province const* capital = nullptr;
+	Culture const* primary_culture = nullptr;
+	Religion const* religion = nullptr;
+	GovernmentType const* government_type = nullptr;
+	NationalValue const* national_value = nullptr;
+	CountryParty const* ruling_party = nullptr;
+	std::vector<Culture const*> accepted_cultures;
+	std::vector<Reform const*> reforms;
+	std::map<Ideology const*, fixed_point_t> upper_house;
+	fixed_point_t plurality = -1, prestige = -1;
+	bool civilised = false;
+	Date last_election = Date(0);
+	Deployment const* initial_oob = nullptr;
+
+	bool updated_accepted_cultures = false, updated_upper_house = false, updated_reforms = false;
+
+	bool ret = expect_dictionary_keys_and_default(
+		[this, &game_manager, &reforms, &updated_reforms, &name](std::string_view key, ast::NodeCPtr value) -> bool {
+			if (game_manager.get_politics_manager().get_issue_manager().has_reform_group_identifier(key)) {
+				updated_reforms = true;
+
+				Reform const* reform;
+
+				bool ret = game_manager.get_politics_manager().get_issue_manager().expect_reform_identifier(assign_variable_callback_pointer(reform))(value);
+				if (std::find(reforms.begin(), reforms.end(), reform) != reforms.end()) {
+					Logger::error("Redefinition of reform ", reform->get_identifier(), " in history of ", name);
+					return false;
+				}
+
+				reforms.push_back(reform);
+				return ret;
+			}
+			// TODO: technologies & inventions
+			return true;
+		},
+		/* we have to use a lambda, assign_variable_callback_pointer apparently doesn't play nice with const & non-const accessors */
```

Comment:
Ah ok yeah `Map::expect_province_identifier` is the only instance of that type of function with a non-const version as well as the usual const one, so the compiler was having difficulty telling which to use, weird. I'll look into it, maybe change the non-const one's name.